edb/server/server.py

```diff
@@ -195,19 +201,34 @@ def get_pg_dbname(self, dbname: str) -> str:
 
     def on_binary_client_connected(self) -> str:
         self._binary_proto_id_counter += 1
+        self._binary_proto_num_connections += 1
+
+        if self._auto_shutdown_handler:
+            self._auto_shutdown_handler.cancel()
+            self._auto_shutdown_handler = None
+
         return str(self._binary_proto_id_counter)
 
     def on_binary_client_authed(self):
-        self._binary_proto_num_connections += 1
         self._report_connections(event='opened')
 
     def on_binary_client_disconnected(self):
         self._binary_proto_num_connections -= 1
         self._report_connections(event="closed")
+
         if not self._binary_proto_num_connections and self._auto_shutdown:
             self._accepting_connections = False
             raise SystemExit
 
+        if not self._binary_proto_num_connections and self._auto_shutdown_after >= 0:
+
+            def shutdown():
+                self._accepting_connections = False
+                raise SystemExit
```

Comment:
@1st1, @fantix Is there a better way other than raising `SystemExit`?

Reply:
A better way would be to have a special "server_done" future object and to do `await asyncio.wait(server.server_forever(), server_done)` in main.py.

Reply:
Actually, since we're talking about auto-shutdown for CI and there are no open connections to EdgeDB, I think this is OK. Just add an XXX, once we have facilities to exit in a better fashion we'll fix this.

Reply:
added XXX comment